forge/src/executor/inspector/cheatcodes/env.rs

```diff
@@ -0,0 +1,171 @@
+use std::collections::BTreeMap;
+
+use super::Cheatcodes;
+use crate::abi::HEVMCalls;
+use bytes::Bytes;
+use ethers::{
+    abi::{AbiEncode, Token, Tokenize},
+    types::{Address, H256, U256},
+    utils::keccak256,
+};
+use revm::{Database, EVMData};
+
+#[derive(Clone, Debug, Default)]
+pub struct Prank {
+    /// Address of the contract that initiated the prank
+    pub prank_caller: Address,
+    /// Address of `tx.origin` when the prank was initiated
+    pub prank_origin: Address,
+    /// The address to assign to `msg.sender`
+    pub new_caller: Address,
+    /// The address to assign to `tx.origin`
+    pub new_origin: Option<Address>,
+    /// The depth at which the prank was called
+    pub depth: u64,
+    /// Whether or not the prank stops by itself after the next call
+    pub single_call: bool,
+}
+
+fn prank(
+    state: &mut Cheatcodes,
+    prank_caller: Address,
+    prank_origin: Address,
+    new_caller: Address,
+    new_origin: Option<Address>,
+    depth: u64,
+    single_call: bool,
+) -> Result<Bytes, Bytes> {
+    let prank = Prank {
+        prank_caller,
+        prank_origin,
+        new_caller,
+        new_origin,
+        // Note: When calling a cheatcode, the depth is not increased. Because of that, the depth
+        // we actually want to prank at is 1 level deeper.
+        depth: depth + 1,
+        single_call,
+    };
+
+    if state.prank.is_some() {
+        return Err("You have an active prank already.".to_string().encode().into())
+    }
+
+    state.prank = Some(prank);
+    Ok(Bytes::new())
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct RecordAccess {
+    pub reads: BTreeMap<Address, Vec<U256>>,
+    pub writes: BTreeMap<Address, Vec<U256>>,
+}
+
+fn record(state: &mut Cheatcodes) {
+    state.accesses = Some(Default::default());
+}
+
+fn accesses(state: &mut Cheatcodes, address: Address) -> Bytes {
+    if let Some(storage_accesses) = &mut state.accesses {
+        ethers::abi::encode(&[
+            storage_accesses.reads.remove(&address).unwrap_or_default().into_tokens()[0].clone(),
+            storage_accesses.writes.remove(&address).unwrap_or_default().into_tokens()[0].clone(),
+        ])
+        .into()
+    } else {
+        ethers::abi::encode(&[Token::Array(vec![]), Token::Array(vec![])]).into()
+    }
+}
+
+pub fn apply<DB: Database>(
+    state: &mut Cheatcodes,
+    data: &mut EVMData<'_, DB>,
+    caller: Address,
+    call: &HEVMCalls,
+) -> Option<Result<Bytes, Bytes>> {
+    Some(match call {
+        HEVMCalls::Warp(inner) => {
+            data.env.block.timestamp = inner.0;
+            Ok(Bytes::new())
+        }
+        HEVMCalls::Roll(inner) => {
+            data.env.block.number = inner.0;
+            // TODO: Set blockhash somehow
```

Comment:
Let's set it as the hash of the block number? Something to consider here is that this would break when using forking tests and roll (the hash wouldn't be the one from mainnet). I don't think we need to do anything about this, but maybe worth documenting

Reply:
Yeah, the main reason I've delayed it until now is that it requires us to create our own database or upstream a fix in revm (which I am going to do) since inspectors have no control over it